xrtl/gfx/render_state.h

```diff
@@ -0,0 +1,629 @@
+// limitations under the License.
+
+#ifndef XRTL_GFX_RENDER_STATE_H_
+#define XRTL_GFX_RENDER_STATE_H_
+
+#include <cstdint>
+
+#include "xrtl/base/fixed_vector.h"
+#include "xrtl/gfx/pixel_format.h"
+
+namespace xrtl {
+namespace gfx {
+
+// Maximum number of vertex bindings and attributes.
+// There may be a larger number supported by the device but this is all the
+// space we reserve for now.
+constexpr int kMaxVertexInputs = 16;
+
+// Maximum number of color attachments.
+// There may be a larger number supported by the device but this is all the
+// space we reserve for now.
+constexpr int kMaxColorAttachments = 8;
+
+enum class PrimitiveTopology {
+  kPointList = 0,
+  kLineList = 1,
+  kLineStrip = 2,
+  kTriangleList = 3,
+  kTriangleStrip = 4,
+  kTriangleFan = 5,
+  kLineListWithAdjacency = 6,
+  kLineStripWithAdjacency = 7,
+  kTriangleListWithAdjacency = 8,
+  kTriangleStripWithAdjacency = 9,
+  kPatchList = 10,
+};
+constexpr int kPrimitiveTopologyBits = 4;
+
+enum class PolygonMode {
+  kFill = 0,
+  kLine = 1,
+  kPoint = 2,
+};
+constexpr int kPolygonModeBits = 2;
+
+enum class CullMode {
+  kNone = 0,
+  kFront = 1,
+  kBack = 2,
+  kFrontAndBack = 3,
+};
+constexpr int kCullModeBits = 3;
+
+enum class FrontFace {
+  kCounterClockwise = 0,
+  kClockwise = 1,
+};
+constexpr int kFrontFaceBits = 1;
+
+// Sample count used for multisampling.
+enum class SampleCount : uint32_t {
+  k1 = 1,
+  k2 = 2,
+  k4 = 4,
+  k8 = 8,
+  k16 = 16,
+  k32 = 32,
+  k64 = 64,
+};
+XRTL_BITMASK(SampleCount);
+constexpr int kSampleCountBits = 7;
+
+enum class BlendFactor {
+  // Color: (0,0,0)
+  // Alpha: 0
+  kZero = 0,
+  // Color: (1,1,1)
+  // Alpha: 1
+  kOne = 1,
+  // Color: (Rs0,Gs0,Bs0)
+  // Alpha: As0
+  kSrcColor = 2,
+  // Color: (1-Rs0,1-Gs0,1-Bs0)
+  // Alpha: 1-As0
+  kOneMinusSrcColor = 3,
+  // Color: (Rd,Gd,Bd)
+  // Alpha: Ad
+  kDstColor = 4,
+  // Color: (1-Rd,1-Gd,1-Bd)
+  // Alpha: 1-Ad
+  kOneMinusDstColor = 5,
+  // Color: (As0,As0,As0)
+  // Alpha: As0
+  kSrcAlpha = 6,
+  // Color: (1-As0,1-As0,1-As0)
+  // Alpha: 1-As0
+  kOneMinusSrcAlpha = 7,
+  // Color: (Ad,Ad,Ad)
+  // Alpha: Ad
+  kDstAlpha = 8,
+  // Color: (1-Ad,1-Ad,1-Ad)
+  // Alpha: 1-Ad
+  kOneMinusDstAlpha = 9,
+  // Color: (Rc,Gc,Bc)
+  // Alpha: Ac
+  kConstantColor = 10,
+  // Color: (1-Rc,1-Gc,1-Bc)
+  // Alpha: 1-Ac
+  kOneMinusConstantColor = 11,
+  // Color: (Ac,Ac,Ac)
+  // Alpha: Ac
+  kConstantAlpha = 12,
+  // Color: (1-Ac,1-Ac,1-Ac)
+  // Alpha: 1-Ac
+  kOneMinusConstantAlpha = 13,
+  // Color: (f,f,f); f = min(As0,1-Ad)
+  // Alpha: 1
+  kSrcAlphaSaturate = 14,
+  // Color: (Rs1,Gs1,Bs1)
+  // Alpha: As1
+  kSrc1Color = 15,
+  // Color: (1-Rs1,1-Gs1,1-Bs1)
+  // Alpha: 1-As1
+  kOneMinusSrc1Color = 16,
+  // Color: (As1,As1,As1)
+  // Alpha: As1
+  kSrc1Alpha = 17,
+  // Color: (1-As1,1-As1,1-As1)
+  // Alpha: 1-As1
+  kOneMinusSrc1Alpha = 18,
+};
+constexpr int kBlendFactorBits = 5;
+
+enum class BlendOp {
+  // R = Rs0 × Sr + Rd × Dr
```

Comment:
nit: could you document the abbreviation scheme here? (uppercase/lowercase RGBA and S?)